### PR TITLE
Stake command to control snapshots flag

### PR DESCRIPTION
**Pull request by vzotova**
Refs #2210

### CODE REVIEW

File: docs/source/guides/network_node/staking_guide.rst [hunk 1]
```diff
@@ -385,6 +387,23 @@ Wind down is *disabled* by default. To start winding down an existing stake:
     (nucypher)$ nucypher stake winddown --hw-wallet
 
 
+Snapshots
+*********
+
+Taking snapshots is *enabled* by default. This option gives ability to be involved in DAO. To stop taking snapshots:
```
Comment (derekpierre) [inline docs/source/guides/network_node/staking_guide.rst:393]:
```suggestion
Taking snapshots is *enabled* by default. Snapshots must be enabled to participate in the DAO. To stop taking snapshots:
```

File: nucypher/blockchain/eth/actors.py [hunk 2]
```diff
@@ -1244,6 +1244,28 @@ def disable_winding_down(self) -> TxReceipt:
         receipt = self._set_winding_down(value=False)
         return receipt
 
+    @property
+    def is_taking_snapshots(self) -> bool:
+        winding_down = self.staking_agent.is_taking_snapshots(staker_address=self.checksum_address)
+        return winding_down
```
Comment (derekpierre) [inline nucypher/blockchain/eth/actors.py:1250]:
```suggestion
        taking_snapshots = self.staking_agent.is_taking_snapshots(staker_address=self.checksum_address)
        return taking_snapshots
```

File: nucypher/cli/literature.py [hunk 3]
```diff
@@ -191,6 +191,22 @@
 SUCCESSFUL_DISABLE_RESTAKING = 'Successfully disabled re-staking for {staking_address}'
 
 
+#
+# Snapshots
+#
+SNAPSHOTS_DISABLING_AGREEMENT = """
+By disabling the taking snapshots for {staking_address}, specified address will be excluded in all future DAO validations
+until snapshots will be enabled again.
```
Comment (derekpierre) [inline nucypher/cli/literature.py:199]:
```suggestion
By disabling snapshots, staker {staking_address} will be excluded from all future DAO validations
until snapshots are enabled.
```

File: nucypher/cli/literature.py [hunk 4]
```diff
@@ -191,6 +191,22 @@
...
+"""
+
+CONFIRM_ENABLE_SNAPSHOTS = "Confirm enable automatic taking snapshots for staker {staking_address}?"
```
Comment (derekpierre) [inline nucypher/cli/literature.py:202]:
```suggestion
CONFIRM_ENABLE_SNAPSHOTS = "Confirm enable automatic snapshots for staker {staking_address}?"
```

File: nucypher/cli/literature.py [hunk 5]
```diff
@@ -191,6 +191,22 @@
...
+
+SUCCESSFUL_ENABLE_SNAPSHOTS = 'Successfully enabled taking snapshots for {staking_address}'
```
Comment (derekpierre) [inline nucypher/cli/literature.py:204]:
```suggestion
SUCCESSFUL_ENABLE_SNAPSHOTS = 'Successfully enabled snapshots for staker {staking_address}'
```

File: nucypher/cli/literature.py [hunk 6]
```diff
@@ -191,6 +191,22 @@
...
+
+CONFIRM_DISABLE_SNAPSHOTS = "Confirm disable taking snapshots for staker {staking_address}?"
```
Comment (derekpierre) [inline nucypher/cli/literature.py:206]:
```suggestion
CONFIRM_DISABLE_SNAPSHOTS = "Confirm disable snapshots for staker {staking_address}?"
```

File: nucypher/cli/literature.py [hunk 7]
```diff
@@ -191,6 +191,22 @@
...
+
+SUCCESSFUL_DISABLE_SNAPSHOTS = 'Successfully disabled taking snapshots for {staking_address}'
```
Comment (derekpierre) [inline nucypher/cli/literature.py:208]:
```suggestion
SUCCESSFUL_DISABLE_SNAPSHOTS = 'Successfully disabled snapshots for staker {staking_address}'
```

File: tests/constants.py [hunk 8]
```diff
@@ -98,7 +98,7 @@
 
 MOCK_INDIVIDUAL_ALLOCATION_FILEPATH = BASE_TEMP_DIR / f'{BASE_TEMP_PREFIX}test-individual-allocation-{datetime.now().strftime(DATETIME_FORMAT)}.json'
 
-MOCK_CUSTOM_INSTALLATION_PATH_2 = '/tmp/nucypher-tmp-test-custom-2-{}'.format(time.time())
+MOCK_CUSTOM_INSTALLATION_PATH_2 = BASE_TEMP_DIR / f'{BASE_TEMP_PREFIX}test-custom-2-{datetime.now().strftime(DATETIME_FORMAT)}'
```
Comment (derekpierre) [inline tests/constants.py:101]: 👍

File: docs/source/guides/network_node/staking_guide.rst [hunk 9]
```diff
@@ -385,6 +387,23 @@ Wind down is *disabled* by default. To start winding down an existing stake:
...
+
+.. code:: bash
+
+    (nucypher)$ nucypher stake snapshots --disable
```
Comment (derekpierre) [inline docs/source/guides/network_node/staking_guide.rst:397]: @vzotova is there a reason someone would want to disable snapshots?

Comment (vzotova) [reply docs/source/guides/network_node/staking_guide.rst:397]: Less gas spending on `mint` and `commitToNextPeriod`, iirc 25k gas less

Comment (derekpierre) [reply docs/source/guides/network_node/staking_guide.rst:397]: 👍 - make sense. If you don't care about participating in the DAO, you save on some gas costs.

**Review by derekpierre — APPROVED** [review]
🎸

File: nucypher/blockchain/eth/actors.py [hunk 10]
```diff
@@ -1244,6 +1244,28 @@ def disable_winding_down(self) -> TxReceipt:
...
 
+    @property
+    def is_taking_snapshots(self) -> bool:
+        taking_snapshots = self.staking_agent.is_taking_snapshots(staker_address=self.checksum_address)
+        return taking_snapshots
+
+    @only_me
+    @save_receipt
+    def _set_snapshots(self, value: bool) -> TxReceipt:
+        # if self.is_contract:
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py:1255]: What are you thinking here?

File: nucypher/cli/literature.py [hunk 11]
```diff
@@ -191,6 +191,22 @@
...
 
 
+#
+# Snapshots
+#
+SNAPSHOTS_DISABLING_AGREEMENT = """
+By disabling snapshots, staker {staking_address} will be excluded from all future DAO validations
+until snapshots are enabled.
+"""
+
```
Comment (KPrasch) [inline nucypher/cli/literature.py:201]: :+1:

File: tests/constants.py (hunk 8 above)
Comment (KPrasch) [inline tests/constants.py:101]: :+1:

**Review by KPrasch — APPROVED** [review]
Just a question about the commented out block.... 

As usual - this is a very thorough PR, even for a smaller changeset!

:cowboy_hat_face:

File: nucypher/blockchain/eth/actors.py (hunk 10 above)
Comment (vzotova) [reply nucypher/blockchain/eth/actors.py:1256]: Placeholder for staking contract interaction, added TODOs

**Review by tuxxy — APPROVED** [review]
Short n sweet PR. :)

File: docs/source/guides/network_node/staking_guide.rst [hunk 12]
```diff
@@ -385,6 +387,23 @@ Wind down is *disabled* by default. To start winding down an existing stake:
     (nucypher)$ nucypher stake winddown --hw-wallet
 
 
+Snapshots
+*********
+
+Taking snapshots is *enabled* by default. Snapshots must be enabled to participate in the DAO. To stop taking snapshots:
```
Comment (cygnusv) [inline docs/source/guides/network_node/staking_guide.rst:393]: Maybe mention why a staker may be interested in disabling snapshots, which is basically to have a slight reduction on gas expenditure.

**Review by cygnusv — APPROVED** [review]

File: docs/source/guides/network_node/staking_guide.rst (hunk 12 above)
Comment (cygnusv) [reply docs/source/guides/network_node/staking_guide.rst:393]:
```suggestion
Taking snapshots is *enabled* by default. Snapshots must be enabled to participate in the DAO, but it has a slight cost in gas every time your staking balance changes. To stop taking snapshots:
```

Comment (vzotova) [reply docs/source/guides/network_node/staking_guide.rst:393]: Ok, I'll add today but later

**Pull request merged**